src/AdapterInterface.php

```diff
@@ -184,6 +184,35 @@ public function write($fileDescriptor, $data, $length, $offset);
      */
     public function close($fd);
 
+    /**
+     * Reads the entire file.
+     *
+     * This is an optimization for adapters which can optimize
+     * the open -> write -> close sequence into one call.
+     *
+     * @param string $path
+     * @param int $offset
+     * @param int|null $length
+     * @return PromiseInterface
+     */
+    public function getContents($path, $offset = 0, $length = null);
+
+    /**
+     * Writes directly to the file.
+     *
+     * This is an optimization for adapters which can optimize
+     * the open -> write -> close sequence into one call.
+     *
+     * Supported flags depend on the adapter implementation,
+     * but may include `LOCK_EX` and `FILE_APPEND`.
```

Comment:
If they depend on the adapter implementation, then I'm not sure how consumers can rely on certain flags to be present? My vote would be to remove any "flags" for now because I see some very good use cases for the underlying feature request and don't want to block this any longer. What do you think? :+1:

Reply:
@clue Users of the low-level adapter API are aware (or at least should be) which adapter they use. So properly documenting what each adapter can do, should be sufficient. Since it's a low-level API, its implementation deeply depends on the actual adapter, and since not every adapter has the same capabilities (e.g. child process can do an exclusive lock and libeio can not), we should still strive for the greatest coverage of an adapter with an as generalized interface as possible.

Flags allow us to do that, and we can still document that every adapter has to support `FILE_APPEND` at the very least (which should be the case, as it allows us to do write and append with one call). Removing flags will lead to custom adapter methods or users of the adapters will need to implement custom userland code to do what the adapter can do, but our adapter API can not provide. Remember that we can do an exclusive lock and append to a file with the same call using the child process adapter, which we can not do with a single call without flags, since an adapter API method can not be added without it being custom to the adapter itself.

My opinion is that the low-level adapter can differ in their implementation, as long as the higher level API (in our case the filesystem and node API) provides the same functionality across all adapters at no cost.